core/src/main/scala/org/typelevel/otel4s/trace/SpanFinalizer.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2022 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s
+package trace
+
+import cats.data.NonEmptyList
+import cats.effect.Resource
+
+sealed trait SpanFinalizer extends Product with Serializable
+
+object SpanFinalizer {
+
+  type Strategy = PartialFunction[Resource.ExitCase, SpanFinalizer]
+
+  object Strategy {
+    def empty: Strategy = PartialFunction.empty
+
+    def reportAbnormal: Strategy = {
+      case Resource.ExitCase.Errored(e) =>
+        SpanFinalizer.multiple(
+          SpanFinalizer.RecordException(e),
+          SpanFinalizer.SetStatus(Status.Error, None)
+        )
+
+      case Resource.ExitCase.Canceled =>
+        SpanFinalizer.SetStatus(Status.Error, Some("canceled"))
+    }
+  }
+
+  final case class RecordException(throwable: Throwable) extends SpanFinalizer
```

Comment:
Should the case classes be hidden? Is it less painful in terms of bincompat? 

```scala
private final case class RecordException(...) extends SpanFinalizer

def recordException(throwable: Throwable): SpanFinalizer = 
  RecordException(throwable)
```

Reply:
If you avoid the `case` and make the constructor private it should be fine. Unless in the future you may want to remove the `RecordException` type, then better not to expose it.